### PR TITLE
mdx: implement MDX generation from generated JSON

### PR DESCRIPTION
**Pull request by imballinst**
This PR implements MDX generation from the generated `doc.json`. To work with this:

1. In the root folder, do `make gen` (if not generated previously)
2. Then, still in the root folder, do `make -C mdx build`. This will transform all `.ts` files to `.js` files in the `mdx/dist` folder.
3. Then, do `make -C mdx gen-mdx`. This will emit the required `.mdx` files to `website/docs` and `.json` dictionary to `website/plugins/proto-messages`.
4. In `website`, install deps using `yarn` (if not installed previously)
5. `yarn start`, then go to http://localhost:3000/docs/booking-v1

## Things that are still in doubt/tentative

1. How to present the "category" name? At the moment we are using the "directory to the package", but all slashes are replaced with dashes. For example, `/booking/v1` are transformed to `booking-v1`.
2. Ideas regarding the TODOs, most importantly these ones (the proto messages plugin):

```
// TODO(imballinst): this plugin has 2 known weaknesses:
// 1. if there is a message with the same name with another "local message",
// then it will link to a wrong link.
// 2. there is no way yet to recognize common know messages (e.g. from Google protobuf).
```

## Demo video

https://user-images.githubusercontent.com/7077157/151301603-58442091-3749-4ebf-b5dc-afa3c759885a.mov

Signed-off-by: Try Ajitiono <ballinst@gmail.com>

### CODE REVIEW

**Comment by imballinst** [comment]
@dio should I also add the commands to generate the MDX to the GitHub actions in this PR?

**Comment by dio** [comment]
> How to present the "category" name? At the moment we are using the "directory to the package", but all slashes are replaced with dashes. For example, /booking/v1 are transformed to booking-v1.


This should follow the package name, 

```
package booking.v1;
```

**Comment by dio** [comment]
> @dio should I also add the commands to generate the MDX to the GitHub actions in this PR?

Sure. Let me check on my PR first.

**Comment by imballinst** [comment]
> > How to present the "category" name? At the moment we are using the "directory to the package", but all slashes are replaced with dashes. For example, /booking/v1 are transformed to booking-v1.
> 
> This should follow the package name,
> 
> ```
> package booking.v1;
> ```

OK! Let me see if that works (not sure if periods are allowed for `.mdx` files in Docusaurus). I think it should.

File: mdx/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# yarn should be already installed from parent Makefile.
```
Comment (dio) [inline mdx/Makefile:1]: Let's put this in the main Makefile.

Comment (dio) [reply mdx/Makefile:1]: But I can do that as a follow up PR.

Comment (imballinst) [reply mdx/Makefile:1]: Sure! Let's do that in the next PR. Thanks!

**Comment by imballinst** [comment]
Yep, that `booking.v1` works! I have updated the PR.

image

**Review by dio — APPROVED** [review]
🎉

**Comment by dio** [comment]
re: there is no way yet to recognize common know messages (e.g. from Google protobuf).

I think we need to recursively generate the deps.

**Comment by imballinst** [comment]
> re: there is no way yet to recognize common know messages (e.g. from Google protobuf).
> 
> I think we need to recursively generate the deps.

Right. I assume they are going to be generated in the `.proto` to `.json` step? Or is it after the JSON is generated?

**Comment by dio** [comment]
@imballinst https://github.com/imballinst/protosaurus/pull/5

**Pull request merged**